File: leetcode2/AllOoneDataStructure/a.cpp
/* Copyright (C) 2020 by iamslash */

#include <cstdio>
#include <list>
#include <unordered_map>
#include <unordered_set>
#include <string>

// 92ms 30.94% 26.6MB 55.56%
// hash map, bidirectional indexing
// O(1) O(N)
struct Bucket {
  int val;
  std::unordered_set<std::string> keys;
};
class AllOne {
 private:
  std::list<Bucket> buckets;
  std::unordered_map<std::string,
                     std::list<Bucket>::iterator> bucketByKey;
 public:
  void inc(std::string key) {
    // 0. enforce node of key
    if (bucketByKey.count(key) == 0)
      bucketByKey[key] = buckets.insert(buckets.begin(), {0, {key}});
    // 1. insert node and update the lookup
    auto u = bucketByKey[key], n = std::next(u);
    if (n == buckets.end() || u->val + 1 < n->val)
      n = buckets.insert(n, {u->val + 1, {}});
    n->keys.insert(key);
    bucketByKey[key] = n;
    // 2. remove node
    u->keys.erase(key);
    if (u->keys.empty())
      buckets.erase(u);
  }
    
  void dec(std::string key) {
    // 0. ensure key node
    if (bucketByKey.count(key) == 0)
      return;
    // 1. insert node and update the lookup
    auto u = bucketByKey[key], p = std::prev(u);
    bucketByKey.erase(key);
    if (u->val > 1) {
      if (u == buckets.begin() || p->val < u->val - 1)
        p = buckets.insert(u, {u->val-1, {}});
      p->keys.insert(key);
      bucketByKey[key] = p;
    }
    // 2. remove node
    u->keys.erase(key);
    if (u->keys.empty())
      buckets.erase(u);
  }
    
  std::string getMaxKey() {
    if (buckets.empty())
      return "";
    return *(buckets.rbegin()->keys.begin());
  }
    
  std::string getMinKey() {
    if (buckets.empty())
      return "";
    return *(buckets.begin()->keys.begin());
  }
};

File: leetcode2/AllOoneDataStructure/a.py
# -*- coding: utf-8 -*-
# Copyright (C) 2023 by iamslash 

# 173ms 85.95% 33.2MB 34.83%
# bidirectional indexing, doubly linked list
class Node:
    def __init__(self, key_set, count, prev=None, next=None):
        self.key_set = key_set
        self.count = count
        self.next = next
        self.prev = prev

class AllOne:
    def __init__(self):
        self.head = Node({""}, count=0)
        self.tail = Node({""}, count=float('inf'), prev=self.head)
        self.head.next = self.tail
        self.key_node_dict = {}

    def inc(self, key: str) -> None:
        if key in self.key_node_dict:
            node = self.key_node_dict[key]
            inc_cnt = node.count + 1
            if len(node.key_set) == 1:
                if inc_cnt == node.next.count:
                    node.next.key_set.add(key)
                    self.key_node_dict[key] = node.next
                    node.next.prev, node.prev.next = node.prev, node.next
                else:
                    node.count = inc_cnt
            else:
                node.key_set.remove(key)
                if inc_cnt == node.next.count:
                    node.next.key_set.add(key)
                    self.key_node_dict[key] = node.next
                else:
                    new_node = Node(key_set={key}, count=inc_cnt, prev=node, next=node.next)
                    node.next.prev = new_node
                    node.next = new_node
                    self.key_node_dict[key] = new_node
        else:
            if self.head.next.count == 1:
                self.head.next.key_set.add(key)
                self.key_node_dict[key] = self.head.next
            else:
                node = Node(key_set={key}, count=1, prev=self.head, next=self.head.next)
                self.head.next.prev = node
                self.head.next = node
                self.key_node_dict[key] = node

    def dec(self, key: str) -> None:
        node = self.key_node_dict[key]
        if node.count == 1:
            if len(node.key_set) == 1:
                node.next.prev, node.prev.next = node.prev, node.next
            else:
                node.key_set.remove(key)
            del self.key_node_dict[key]
        else:
            dec_cnt = node.count - 1
            if len(node.key_set) == 1:
                if dec_cnt == node.prev.count:
                    node.prev.key_set.add(key)
                    self.key_node_dict[key] = node.prev
                    node.next.prev, node.prev.next = node.prev, node.next
                else:
                    node.count = dec_cnt
            else:
                node.key_set.remove(key)
                if dec_cnt == node.prev.count:
                    node.prev.key_set.add(key)
                    self.key_node_dict[key] = node.prev
                else:
                    new_node = Node(key_set={key}, count=dec_cnt, prev=node.prev, next=node)
                    node.prev.next = new_node
                    node.prev = new_node
                    self.key_node_dict[key] = new_node

    def getMaxKey(self) -> str:
        return next(iter(self.tail.prev.key_set))

    def getMinKey(self) -> str:
        return next(iter(self.head.next.key_set))
